File: leet-code/738.单调递增的数字.js
/* 🔖
 * @lc app=leetcode.cn id=738 lang=javascript
 *
 * [738] 单调递增的数字
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var monotoneIncreasingDigits = function (n) {
  // 从右往左遍历，发现非递减则当前位 -1 ，往后低位补充 9
  // 不能从左往右，如 122045，会导致 2 减 1 后变成 121999还是不符合
  let arr = n.toString().split('')
  let fillIdx = -1
  for (let i = arr.length - 2; i >= 0; i -= 1) {
    if (arr[i] > arr[i + 1]) {
      // 非递减，不用考虑 arr[i-1] 的情况，因为下一个循环会自动比较去减
      arr[i] -= 1
      fillIdx = i + 1
    }
  }

  if (fillIdx !== -1) {
    arr.fill(9, fillIdx)
  }

  return +arr.join("")
};
// @lc code=end


// @after-stub-for-debug-begin
module.exports = monotoneIncreasingDigits;
// @after-stub-for-debug-end

File: leet-code-rs/src/solutions/738_单调递增的数字.rs
/*
 * @lc app=leetcode.cn id=738 lang=rust
 *
 * [738] 单调递增的数字
 */

// @lc code=start

#[allow(unused)]
impl Solution {
    pub fn monotone_increasing_digits(n: i32) -> i32 {
        let mut str = n.to_string().chars().collect::<Vec<char>>();

        // 从右到左匹配，如果发现左边比右边大，则左边数字-1，并把后面位开始位置 变为 9。否则保持原来数字

        let mut idx = None;
        for i in (0..str.len() - 1).rev() {
            if str[i] > str[i + 1] {
                // 不是递增
                str[i] = (str[i] as u8 - 1) as char;
                // idx 往后填充 9
                idx = Some(i + 1)
            }
        }

        if let Some(idx) = idx {
            str[idx..].fill('9');
        }

        return str.into_iter().collect::<String>().parse().unwrap();
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_monotone_increasing_digits() {
        //assert_eq!(Solution::monotone_increasing_digits(vec![]),[]);
        assert!(true)
    }
}
